src/main/java/org/wildfly/security/keystore/KeyStoreWatcher.java

```diff
@@ -0,0 +1,227 @@
+
+        File parentDir = canonical.getParentFile();
+        String fileName = canonical.getName();
+
+        Path dirPath = parentDir.toPath();
+
+        boolean watchRequired = false;
+        Map<String, List<Store>> pathRegistration = watchedPaths.get(dirPath);
+        List<Store> pathStores = null;
+        if (pathRegistration != null) {
+            pathRegistration = new HashMap<String, List<Store>>(pathRegistration);
+            List<Store> tmpStores = pathRegistration.get(fileName);
+            if (tmpStores != null) {
+                pathStores = new LinkedList<Store>(tmpStores);
+                Iterator<Store> storeIterator = pathStores.iterator();
+                while (storeIterator.hasNext()) {
+                    Store current = storeIterator.next();
+                    if (keyStore == current) {
+                        storeIterator.remove();
+                    }
+                }
+                if (pathStores.isEmpty()) {
+                    pathRegistration.remove(fileName);
+                } else {
+                    pathRegistration.put(fileName, pathStores);
+                }
+            }
+
+            Map<Path, Map<String, List<Store>>> newWatchedPaths = new HashMap<Path, Map<String, List<Store>>>(watchedPaths);
+            if (pathRegistration.isEmpty()) {
+                newWatchedPaths.remove(dirPath);
+                WatchKey key = registrations.remove(dirPath);
+                if (key != null) {
+                    key.cancel();
+                }
+                if (newWatchedPaths.isEmpty()) {
+                    watchService.close();
+                    watchService = null;
+                }
+            } else {
+                newWatchedPaths.put(dirPath, pathRegistration);
+            }
+            watchedPaths = newWatchedPaths;
+        }
+
+    }
+
+    interface Store {
+
+        void modified();
+
+    }
+
+    private class EventTaker implements Runnable {
+
+        @Override
+        public void run() {
+            try {
+                while (watchService != null) {
+                    WatchKey key = watchService.take();
+                    Path watchedPath = (Path) key.watchable();
+                    Map<String, List<Store>> pathRegistration = watchedPaths.get(watchedPath);
+                    if (pathRegistration != null) {
+                        for (WatchEvent<?> event : key.pollEvents()) {
+                            if (StandardWatchEventKinds.ENTRY_CREATE.equals(event.kind())
+                                    || StandardWatchEventKinds.ENTRY_MODIFY.equals(event.kind())) {
+                                Path context = (Path) event.context();
+                                String name = context.getFileName().toString();
+                                List<Store> stores = pathRegistration.get(name);
+                                if (stores != null) {
+                                    for (Store current : stores) {
```

Comment:
I will double check that but the collection modification should have been to a copy of the collection - that is why I synchronized the two modification methods but left access unsynchronized.